drivers/include/sx127x.h

```diff
@@ -0,0 +1,704 @@
+ * @brief   SX127X initialization result.
+ */
+enum {
+    SX127X_INIT_OK = 0,                /**< Initialization was successful */
+    SX127X_ERR_SPI,                    /**< Failed to initialize SPI bus or CS line */
+    SX127X_ERR_TEST_FAILED,            /**< SX127X testing failed during initialization (check chip) */
+    SX127X_ERR_THREAD                  /**< Unable to create DIO handling thread (check amount of free memory) */
+};
+
+/**
+ * @brief   Radio driver supported modems.
+ */
+enum {
+    SX127X_MODEM_FSK = 0,              /**< FSK modem driver */
+    SX127X_MODEM_LORA,                 /**< LoRa modem driver */
+};
+
+/**
+ * @brief   LoRa signal bandwidth.
+ */
+enum {
+    SX127X_BW_125_KHZ = 0,             /**< 125 kHz bandwidth */
+    SX127X_BW_250_KHZ,                 /**< 250 kHz bandwidth */
+    SX127X_BW_500_KHZ                  /**< 500 kHz bandwidth */
+};
+
+/**
+ * @brief   LoRa spreading factor rate
+ */
+enum {
+    SX127X_SF6 = 6,                    /**< spreading factor 6 */
+    SX127X_SF7,                        /**< spreading factor 7 */
+    SX127X_SF8,                        /**< spreading factor 8 */
+    SX127X_SF9,                        /**< spreading factor 9 */
+    SX127X_SF10,                       /**< spreading factor 10 */
+    SX127X_SF11,                       /**< spreading factor 11 */
+    SX127X_SF12                        /**< spreading factor 12 */
+};
+
+/**
+ * @brief   LoRa error coding rate.
+ */
+enum {
+    SX127X_CR_4_5 = 1,                 /**< coding rate 4/5 */
+    SX127X_CR_4_6,                     /**< coding rate 4/6 */
+    SX127X_CR_4_7,                     /**< coding rate 4/7 */
+    SX127X_CR_4_8                      /**< coding rate 4/8 */
+};
+
+/**
+ * @brief   Radio driver internal state machine states definition.
+ */
+enum {
+    SX127X_RF_IDLE = 0,                /**< Idle state */
+    SX127X_RF_RX_RUNNING,              /**< Sending state */
+    SX127X_RF_TX_RUNNING,              /**< Receiving state */
+    SX127X_RF_CAD,                     /**< Channel activity detection state */
+};
+
+/**
+ * @brief   Event types.
+ */
+enum {
+    SX127X_RX_DONE = 0,                /**< Receiving complete */
+    SX127X_TX_DONE,                    /**< Sending complete*/
+    SX127X_RX_TIMEOUT,                 /**< Receiving timeout */
+    SX127X_TX_TIMEOUT,                 /**< Sending timeout */
+    SX127X_RX_ERROR_CRC,               /**< Receiving CRC error */
+    SX127X_FHSS_CHANGE_CHANNEL,        /**< Channel change */
+    SX127X_CAD_DONE,                   /**< Channel activity detection complete */
+};
+
+/**
+ * @brief   LoRa configuration structure.
+ */
+typedef struct {
+    uint8_t power;                     /**< Signal power */
+    uint8_t bandwidth;                 /**< Signal bandwidth */
+    uint8_t datarate;                  /**< Spreading factor rate, e.g datarate */
+    bool low_datarate_optimize;        /**< Optimize for low data rate */
```

Comment:
I'm not sure I agree.

First of all there is the number of saved memory bytes, which *is* significant in my opinion. Let's look at the current case and an optimized case in comparison. In the current case the memory on a 32-bit platform looks like this

```
0                   1                   2                   3
0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2
+---------------+---------------+---------------+---------------+
| power         | bandwidth     | datarate      |               |   32 bytes
+---------------+---------------+---------------+---------------+
| low_datarate_optimize                                         |   64 bytes
+---------------+---------------+-------------------------------+
| coderate      |               | preamble_len                  |   96 bytes
+---------------+---------------+-------------------------------+
| use_fix_len                                                   |  128 bytes
+---------------------------------------------------------------+
| crc_on                                                        |  160 bytes
+---------------------------------------------------------------+
| freq_hop_on                                                   |  192 bytes
+---------------------------------------------------------------+
| freq_hop_per  |                                               |  224 bytes
+---------------+-----------------------------------------------+
| iq_inverted                                                   |  256 bytes
+---------------------------------------------------------------+
| rx_continuous                                                 |  288 bytes
+---------------------------------------------------------------+
| rx_timeout                                                    |  320 bytes
+---------------------------------------------------------------+
| tx_timeout                                                    |  352 bytes
+---------------------------------------------------------------+
```

Compare that to a reordered struct with flags instead of bools:

```
0                   1                   2                   3
0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2
+---------------+---------------+---------------+---------------+
| power         | bandwidth     | datarate      | coderate      |   32 bytes
+---------------+---------------+-------------------------------+
| preamble_len                  | freq_hop_on   | flags         |   64 bytes
+-------------------------------+-------------------------------+
| rx_timeout                                                    |   96 bytes
+---------------------------------------------------------------+
| tx_timeout                                                    |  128 bytes
+---------------------------------------------------------------+
```

That is a gain of 224 bytes, which is huge in our terms (remember, we support systems that only have kBytes of RAM). Even in the non-reordered case we would still gain 196 bytes! If that doesn't drive my point home: Even on 16-bit systems (and even 8-bit platforms maybe, I don't know exactly how they pack their structs tbh), where we typically have even less RAM we gain 160 bytes with that optimization.

Regarding the flash and execution time we have to distinguish between setting (`x = true` vs `x |= FLAG`), unsetting (`x = false` vs `x &= ~FLAG`) and comparing (`(x)` vs `(x & FLAG)`) the variable/flag since they differ in the flag case 

* For setting the byte we have the same number of ops on ARM. `MOV` for the bool case, `ORR` for the flag case.
* For unsetting ARM seems to have a special operation for bitwise unset (`BICS` [?] at least according to my analysis) so we again have have one op vs one (bool uses MOV again)
* Finally I think we win again some ROM when using flags in the compare case, since the compiler translates the `(x)` check to an `UXTB` and `CMP`, while (x & FLAG) is compiled to a simple bitwise and + compare register update `ANDS`.

All in all, especially since you think the priority for that is low I would argue that we then should do it right the first time, otherwise we run around with unoptimized code for years ;-).